### PR TITLE
MM-32419 Auto send an outgoing webhook on incident archive

### PR DESCRIPTION
**Pull request by hakansa**
#### Summary

Implementation of automated outgoing webhook when an incident archived.

Many thanks to @agarciamontoro for exhaustive help and support.  I've read a lot the PR https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/503  when I was working on this feature.

#### Changes in the Backstage

A new "When an incident archived" Section and Webhook component added.

Screen Shot 2021-04-21 at 22 04 50

Screen Shot 2021-04-21 at 22 05 23

#### Changes in the DB
- New columns in `IR_Playbook`:
    * `WebhookOnArchiveURL`, which contains the, ehem, webhook URL. 
    * `WebhookOnArchiveEnabled`, that stores the on/off state of the setting.
- New column in IR_Incident:
    * `WebhookOnArchiveURL`, which stores a copy of the data in the playbook's column when the incident is created. If the playbook's `WebhookOnArchiveEnabled` contains false, the data is not copied, so there's no need to add `WebhookOnArchiveEnabled` column to the `IR_Incident` table.

#### Tests
I tested that the webhook is sent and that the payload is correct in `server/incident/service_test.go`.
#### Ticket Link
https://mattermost.atlassian.net/browse/MM-32419

### CODE REVIEW

**Comment by agarciamontoro** [comment]
Thank you for your contribution, @hakansa :tada: I'll try to review this today, tomorrow at the latest!

In the meanwhile: I see that the E2E tests are failing, do you know if they're failing also for you locally?

**Comment by hakansa** [comment]
@agarciamontoro I'm not sure why e2e are failing.. I just added `webhookOnArchiveURL` to  `plugin_api_commands.js` and push it but tests are still failing. I'm looking for what the problem is.

**Comment by agarciamontoro** [comment]
I'd suggest to first try if they are failing in your dev environment too, we've been having some problems with the E2E tests in CI and this may be another symptom of that problem.

For running them locally:
1. First, make sure you have your local server running with your version of the plugin installed.
2. Then, `cd` into `tests-e2e` and run `npm i` to install cypress. 
3. Finally, run `npm run cypress:open`. This will open Cypress, from where you can run the test files.

Try to run some of the tests that are failing on CI and see if they fail for you locally.

**Comment by hakansa** [comment]
I'm getting `HTTP 401 Unauthorized` error for every test on Cypress. Should I add an user for cypress  like Username: `sysadmin` ?

**Comment by agarciamontoro** [comment]
We are relying on test data on the server, sorry that I forgot to mention that!

In the mattermost-server repo, run `make test-data`. That should create the needed teams and users on which the E2E tests rely.

**Comment by hakansa** [comment]
Actually I couldn't find any problem on e2e-tests.. Also I double checked my commits but I didn't see any piece of code that could break e2e tests..

**Comment by agarciamontoro** [comment]
Let's see if merging master fixes things!

**Comment by agarciamontoro** [comment]
I'll check this locally tomorrow, E2E tests are always brittle!

**Comment by hakansa** [comment]
@agarciamontoro Thanks! I will also try to look again..

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
Thank you very much for your work, @hakansa! :tada:

I left some suggestions below, mostly typos and some code refactoring issues, but otherwise it looks really good!

I've been taking a look at the E2E tests and, of course, the failures do not reproduce locally. Let's tackle that problem after you can address the changes requested here. I'm not entirely sure what's happening there.

**Comment by hakansa** [comment]
@agarciamontoro hi again, I push new commits by considering your reviews.  Some tests were failed on CI (`test-MySQL57-Postgres10 `) , but these tests passing locally.

**Comment by agarciamontoro** [comment]
Hm, it looks like you found a race condition, but it's actually happening on the other webhook. I cannot reproduce locally wither, but that's what race conditions do to us.

As for this PR, the same issue may affect your code too, let me see.

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
Some extra changes to prevent more race conditions, and I think we're good to go!

**Comment by hakansa** [comment]
@agarciamontoro thanks for suggestions.. I clearly understand what's happening there.. 
I pushed the new commits and it looks like tests will be OK!

File: server/api/playbooks.go [hunk 1]
```diff
@@ -138,6 +138,20 @@ func (h *PlaybookHandler) createPlaybook(w http.ResponseWriter, r *http.Request)
 		}
 	}
 
+	if pbook.WebhookOnArchiveURL != "" {
```
Comment (agarciamontoro) [inline server/api/playbooks.go:141]: QA just found a bug with this on my own implementation of the webhook on creation, and it reproduces here: if the UI toggle is disabled but an invalid URL was entered, the playbook cannot be created nor saved, so we need to change this condition:
```suggestion
	if pbook.WebhookOnArchiveEnabled {
```

File: server/api/playbooks.go [hunk 2]
```diff
@@ -281,6 +295,20 @@ func (h *PlaybookHandler) updatePlaybook(w http.ResponseWriter, r *http.Request)
...
```
Comment (agarciamontoro) [inline server/api/playbooks.go:298]:
```suggestion
	if pbook.WebhookOnArchiveEnabled {
```

**Review by agarciamontoro — CHANGES_REQUESTED** [review]
Amazing, @hakansa, thank you very much! :tada:

QA just found a bug with in my own implementation of the webhook on creation, and it reproduces here: if the UI toggle is disabled but an invalid URL was entered, the playbook cannot be created nor saved, so we need to change a couple of conditions in the API handlers. I left the suggestions below:

**Comment by hakansa** [comment]
@agarciamontoro thanks for review.. I pushed the new commits..

File: server/api/playbooks.go (hunk 2 above)
Comment (agarciamontoro) [reply server/api/playbooks.go:298]: @hakansa: we also need it here, in the updatePlaybook function!

Comment (hakansa) [reply server/api/playbooks.go]: @agarciamontoro oops.. I push a  new commit to fix that.

**Review by agarciamontoro — APPROVED** [review]
Yay! :tada: Thank you very much for all your work on this, @hakansa, it will be really useful!

File: server/sqlstore/migrations.go [hunk 3]
```diff
@@ -630,15 +630,84 @@ var migrations = []Migration{
...
+					return errors.Wrapf(err, "failed adding column WebhookOnArchiveURL to table IR_Incident")
+				}
+
+				if err := addColumnToPGTable(e, "IR_Playbook", "WebhookOnArchiveURL", "TEXT DEFAULT ''"); err != nil {
+					return errors.Wrapf(err, "failed adding column WebhookOnArchiveURL to table IR_Playbook")
+				}
+
+				if err := addColumnToPGTable(e, "IR_Playbook", "WebhookOnArchiveEnabled", "BOOLEAN DEFAULT FALSE"); err != nil {
+					return errors.Wrapf(err, "failed adding column WebhookOnArchiveEnabled to table IR_Playbook")
+				}
```
Comment (cpoile) [inline server/sqlstore/migrations.go:671]: I think these still need to be removed  from the 0.13->0.14, no?

**Review by cpoile — CHANGES_REQUESTED** [review]
Great work and thank you @hakansa !

One small change (I think).

File: server/sqlstore/migrations.go (hunk 3 above)
Comment (hakansa) [reply server/sqlstore/migrations.go:671]: @cpoile  thanks! I'm not sure, the comment below is opposite of that.
https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/526#discussion_r619096213

Comment (cpoile) [reply server/sqlstore/migrations.go:671]: I think that comment is saying the same thing I'm saying -- you shouldn't modify previously existing migrations. You can assume previous migrations have already been run, so whatever you add will not be executed. Put whatever new columns/tables into a new separate migration. 

Make sense? Of course let me know if I'm misreading your code. :)

Comment (hakansa) [reply server/sqlstore/migrations.go]: @cpoile i think i couldn't explain myself properly, sorry for that..

You are absolutetly right. In my previous commits, i updated the previous migration function, and as @agarciamontoro  suggests, i write a new migration function ( from 0.14 to 0.15) and added the new column migrations there.

On the other hand, I forgot to remove my old additions to older migrate function..  I removed that blocks and push the new commits..

Thanks for your review!

Comment (agarciamontoro) [reply server/sqlstore/migrations.go]: Oh I had completely missed that, thank you for the second pair of eyes, @cpoile!

**Review by cpoile — APPROVED** [review]
Awesome, thanks @hakansa !

**Comment by cpoile** [comment]
@agarciamontoro Should we invite Prapti now?

**Comment by agarciamontoro** [comment]
Yes, we should!

**Comment by agarciamontoro** [comment]
Last commit just fixes the conflicts we had with `api/playbooks`

**Comment by prapti** [comment]
In test queue.

**Comment by hakansa** [comment]
@prapti thanks! Are there any updates in here?

**Comment by prapti** [comment]
Hi @hakansa! Sorry it took a while to get to this PR. I think it looks great except for one little clarification that I may need with regards to invalid webhook URL while archiving an incident -- I see that with an invalid URL, there is no notification in the incident channel that the outgoing webhook has failed when archiving the incident. In a similar scenario, the outgoing webhook 'when an incident starts' posts a message in the incident channel that "Incident creation announcement through the outgoing webhook failed. Contact your System Admin for more information." If this is an actual issue, honestly I think taking care of it in a different PR will be fine too.
@agarciamontoro thoughts? Not sure if this has already been discussed.

**Comment by agarciamontoro** [comment]
@prapti, we should definitely have that warning if the request failed, but I see that the code is actually there: https://github.com/mattermost/mattermost-plugin-incident-collaboration/pull/526/files#diff-a3c78b2448c749821fb13b31a33b5b285db50d8f18c8991b165cb0b434434579R746 It doesn't work?

**Comment by prapti** [comment]
@agarciamontoro Yeah, I don't see the 'invalid webhook' message coming through on archiving, only on incident start (seen in the screenshot below). The following was set up using the same invalid webhook `http://badwebhook` for both starting and archiving the incident:
Screen Shot 2021-05-20 at 5 43 06 PM

**Comment by agarciamontoro** [comment]
Thanks for checking @prapti! @hakansa, could you take a look? I don't see any obvious error in the code, but something must be missing :thinking:

**Comment by hakansa** [comment]
@agarciamontoro i've take a look but i couldn't see any obvious code..

**Comment by prapti** [comment]
@hakansa @agarciamontoro Wondering if either of you can reproduce this issue if I had it tested incorrectly. Would you be willing to give this invalid webhook case a try?

File: server/incident/service.go [hunk 4]
```diff
@@ -674,6 +738,17 @@ func (s *ServiceImpl) UpdateStatus(incidentID, userID string, options StatusUpda
 		return err
 	}
 
+	if options.Status == StatusArchived {
+		if incidentToModify.WebhookOnArchiveURL != "" {
+			go func() {
+				if err = s.sendWebhookOnArchive(*incidentToModify); err != nil {
+					s.pluginAPI.Log.Warn("failed to send a POST request to the archive webhook URL", "webhook URL", incidentToModify.WebhookOnArchiveURL, "error", err)
+					_, _ = s.poster.PostMessage(incidentToModify.BroadcastChannelID, "Incident archived announcement through the outgoing webhook failed. Contact your System Admin for more information.")
```
Comment (agarciamontoro) [inline server/incident/service.go:746]: The ID of the incident's channel is stored in the `ChannelID` field. The `BroadcastChannelID` field contains the ID of a channel where the incident creation is optionally broadcasted. That's why we were not seeing the error in the incident's channel:
 
```suggestion
					_, _ = s.poster.PostMessage(incidentToModify.ChannelID, "Incident archived announcement through the outgoing webhook failed. Contact your System Admin for more information.")
```

**Review by agarciamontoro — CHANGES_REQUESTED** [review]

File: server/incident/service.go (hunk 4 above)
Comment (hakansa) [reply server/incident/service.go]: @agarciamontoro thanks for correction! I've pushed a new commit to fix that.

**Comment by agarciamontoro** [comment]
@hakansa, if you can merge master and fix the conflicts, I think we will be able to merge this as soon as @prapti approves it! Thank you for all your effort here :pray:

**Review by agarciamontoro — APPROVED** [review]

**Comment by hakansa** [comment]
@agarciamontoro thanks.. I've fixed some of the conflicts but I have to fix some linting problems etc. I will make a commit a few hours later and fix all of issues.

**Comment by agarciamontoro** [comment]
Great, thank you, @hakansa!

**Comment by agarciamontoro** [comment]
Hi @hakansa! We just merged [a PR](https://github.com/mattermost/mattermost-server/issues/17317) that has created some more conflicts here. Would you mind taking a look at those? Sorry for the additional work!

**Comment by lieut-data** [comment]
@hakansa, don't worry about the merge conflicts -- we'll tackle these while we finish up review. Thanks for your patience :)

**Comment by agarciamontoro** [comment]
Hi again @hakansa! After we merged https://github.com/mattermost/mattermost-server/issues/17317, which sends an outgoing webhook on every status update, we realized its scope already covers the problem solved by your PR, which sends the outgoing webhook on the particular case of the status changing to Archived. After a discussion with the team, we decided that closing this PR is the best decision for the product, as merging it would mean having duplicated code that would need duplicated maintenance effort.

Of course, this is completely on me for not having identified the duplication of efforts before, and I am really really sorry that your hard work is not being merged. Please get in touch here or in community.mattermost.com (I'm @alejandro.garcia there) if you have any doubts, if you want to keep contributing, or simply if you want to have a quick chat with us! :)

By the way, I'm in touch with a colleague to try to get you some swag anyway, because in a normal scenario, we would have merged your PR, its quality is really good and it definitely needs some reward. We'll get back to you :)

**Comment by agarciamontoro** [comment]
Yay, we got you a little something! Please, get in touch with me in https://community.mattermost.com/core/messages/@alejandro.garcia or through my email: alejandro dot garcia at mattermost dot com :)

**Pull request closed without merging**